### DISCUSSION TITLE
How to use a custom reporter

### DISCUSSION THREAD

**Discussion by richard-ling** [remarkjs/remark#1173] [Q&A] [answered]
I would like to use remark-lint to lint markdown files in my repo, both in vscode and on CI.
The default formatting of lint warnings does not enable clicking the error to open the file in vscode.
eg: if I lint my packages/docs folder, the current output is:

    README.md
          22:43-22:94  warning  <Some error message...>

What I want is:

    /absolute/path/to/repo/packages/docs/README.md:22:43 Some error message...

Then I can command-click to open the correct file at the correct line. Currently vscode will sometimes open the wrong file, as there's no absolute path; and it can't go to the correct line.

- Is there an existing reporter that works properly in vscode?  (Seems vfile-reporter-position may work?)
- Is there an example showing how to use a different reporter, in either .remarkrc or on the command line?

As a general comment, I find the whole remark-lint ecosystem very confusing to navigate. It seems like the documentation is very fragmented, with little guidance on how to do simple tasks.  Anyway, hopefully I can get an answer here.  Thanks!

**Comment by wooorm** [comment]
If you use remark with VS Code, I recommend https://github.com/remarkjs/vscode-remark.

As for how to use remark-cli, use `remark --help`.
This includes the line:

```
      --report <reporter>                 specify reporter
```

That means you can do for example `remark --report ./my-reporter.js`. Some more info is linked from the `remark-cli` readme, to https://github.com/unifiedjs/unified-args#--report-reporter.

**Comment by richard-ling** [comment]
Thanks @wooorm !  I somehow missed the --report CLI option.

How can I specify the reporter in the .remarkrc?  I'd prefer to do this in one place, rather than add the CLI argument to ~100 package.jsons.  So I have tried these .remarkrcs, referring to https://github.com/unifiedjs/unified-engine/blob/main/doc/options.md#optionsreporter:

    {
      "plugins": [ ... ],
      "reporter": "position"
    }

    {
      "plugins": [ ... ],
      "settings": {
        "reporter": "position"
      }
    }

    {
      "plugins": [ ... ],
      "report": "position"  (to match the CLI option name)
    }

I know the .reportrc is being used, because the plugins take effect.  But the reporter doesn't.
Thanks again!

**Reply by wooorm** [reply, marked as answer]
No worries!

You cannot specify reporters in `.remarkrc`

Those config files are files on the file system. Different files could have different configuration.
But we can’t use some formatter for this file and some other formatter for some other file.

> CLI argument to ~100 package.jsons

Why do you have 100 `package.json`s?! 🤔🤯

**Reply by richard-ling** [reply]
Yeah it's a lot. It's a large monorepo.

Every package.json has a script `"lint:md": "remark --no-stdout --use remark-lint *.md"`, where I would have to add the `--report` CLI argument. But maybe a better solution is to write a new script using the remark API (something like [this](https://github.com/remarkjs/remark-lint#example-check-markdown-on-the-api)) and call that, rather than calling `remark` directly.  Then in future we just update the script to change its behaviour.

**Reply by wooorm** [reply]
Perhaps you can do a single format in the monorepo root. Typically, running scripts in 100 folders is much slower than running it once from a parent folder.

Example: https://github.com/micromark/micromark-extension-gfm-table/blob/8c8aa31a3d645b8a25f88859941554beca3ebe12/package.json#L65

**Reply by richard-ling** [reply]
Just to follow up on your question (months later), yeah it would be quicker to do it how you suggest.
The justification for doing separately for each package is the different packages are managed by different teams and when a CI job on a package fails, the relevant team is notified. There's a bit of inertia around this practice.
But thanks - it is true that your suggestion would work better.
